feature: add JSON reporter

gijsdb opened the pull request:
This PR addresses issue #11

gijsdb commented:
@kehoecj Let me know if it's how you want it.

kehoecj commented:
@gijsdb Thanks again for the MR. This looks really good! A couple findings from the review:

## Unit Test Coverage
The following lines are being missed in the coverage report:
* [cli.go](https://github.com/Boeing/config-file-validator/pull/18/files#diff-3040c236897f2704958b674ce81c445b6de53f2ff4c204c812ec510de8a76a73R46)
* [json_reporter.go](https://github.com/Boeing/config-file-validator/pull/18/files#diff-d237a8957e30500e8a3c7e9d00f3ebde1fd9d34b3b1253691e0583c5027bc85aR58)

## Output
On windows I'm noticing a double backslash on the JSON output:
![image](https://user-images.githubusercontent.com/118750525/235565795-0fc7d5d9-c447-41be-a833-90961f8dfdeb.png)

gijsdb commented:
@kehoecj 

I've added a test case for the missing line in cli.go. 

Does the test in reporter_test.go not check for that error in json_reporter.go? 

Also I am not on a windows system so I cannot confirm the check and replace for "\" works but I've added a test case for it, which passes.

kehoecj commented:
> @kehoecj
> 
> I've added a test case for the missing line in cli.go.

Confirmed coverage, thanks!

> Does the test in reporter_test.go not check for that error in json_reporter.go?

Yeah, it looks like it's just not hitting the error case for when the call to `json.MarshalIndent` fails. Admittedly this would be pretty rare since the json is being pre-formatted but it's possible if the error message injects some JSON like string in there that it could fail

> Also I am not on a windows system so I cannot confirm the check and replace for "\" works but I've added a test case for it, which passes.

Looks good - this is what I see now:

![image](https://user-images.githubusercontent.com/118750525/235692399-2f97b6d6-d1f7-48d8-a72c-e4f5ee303cd4.png)

gijsdb commented:
@kehoecj I'm unsure how I would go about covering that json.MarshalIndent case, do you have any suggestions? Asserting the error is nil maybe using something like testify?

kehoecj commented:
> @kehoecj I'm unsure how I would go about covering that json.MarshalIndent case, do you have any suggestions? Asserting the error is nil maybe using something like testify?

I looked into this a bit and there is no real way to test this because it's essentially impossible to fail in the current implementation of `json.MarshalIndent`. We control the structure of the object that gets passed into the marshal function. 

I am good with the tests as they are. I'm going to have one more person from my team review this and then we can get it merged in.

kehoecj commented on pkg/cli/cli.go:
```diff
@@ -39,13 +40,20 @@ func Init(searchPath string, excludeDirs []string) CLI {
+
+	if reportType == "json" {
+		reporterType = reporter.JsonReporter{}
+	} else {
```
Might want to throw an error if the report type is not json or standard just so the user is not confused.

gijsdb replied:
I'll add it in validator.go::getFlags where it checks the searchPathPtr is not equal to an empty string. Means I don't have to change the cli.go::Init function to return an error.

jd4235 commented on cmd/validator/validator.go:
```diff
@@ -13,27 +13,31 @@ The flags are:
+func getFlags() (*string, *string, *string, int) {
 	searchPathPtr := flag.String("search-path", "", "The search path for configuration files")
 	excludeDirsPtr := flag.String("exclude-dirs", "", "Subdirectories to exclude when searching for configuration files")
+	reportTypePtr := flag.String("reporter", "standard", "Format of the printed report")
```
Identify all available options in flag description instead of just "Format of the printed report".  For example, "Format of the printed report. Options are `standard` and `json`"

jd4235 commented on pkg/reporter/json_reporter.go:
```diff
@@ -0,0 +1,68 @@
+// Print implements the Reporter interface by outputting
+// the report content to stdout as JSON
+func (jr JsonReporter) Print(reports []Report) error {
+	var reportJSON reportJSON
```
Refrain from making the variable name the same name as the type name.  For example, you could use `report` for the variable name instead of `reportJSON`.

jd4235 commented on pkg/reporter/json_reporter.go:
```diff
@@ -0,0 +1,68 @@
+			errorStr = r.ValidationError.Error()
+		}
+
+		if strings.Contains(r.FilePath, "\\") {
```
I assume this is to convert Windows-style file paths.  Please add comment.

jd4235 reviewed (approved):
Please address comments.

gijsdb commented on cmd/validator/validator.go:
```diff
@@ -13,27 +13,31 @@ The flags are:
 	"log"
 	"os"
 	"strings"
+
```
This is separation between standard packages and other dependencies. Pretty standard in Go programs and automatically enforced by my IDE using official Go formatting.
